File: Grind75/Utility/python/List.py
from __future__ import annotations
from typing import Optional

class ListNode:
  def __init__(self, val=0, nxt=None):
    self.val: int = val
    self.next: ListNode = nxt

def makeList(lst: Optional[list]) -> Optional[ListNode]:
  if not lst:
    return None
  head = ListNode(lst[0])
  node = head
  for i in range(1, len(lst)):
    node.next = ListNode(lst[i])
    node = node.next
  return head

def extractList(lst: Optional[ListNode]) -> list:
  extracted = []
  while lst is not None:
    extracted.append(lst.val)
    lst = lst.next
  return extracted

def printList(lst: Optional[list]) -> None:
  print('[', end='')
  while lst:
    print(lst.val, ', ' if lst.next else '', sep='', end='')
    lst = lst.next
  print(']')

def strList(lst: Optional[list]) -> None:
  result = '['
  while lst:
    result = result + str(lst.val) + (', ' if lst.next else '')
    lst = lst.next
  result = result + ']'
  return result

def compareLists(list1: Optional[ListNode], list2: list) -> bool:
  for v in list2:
    if list1.val != v:
      return False
    list1 = list1.next
  return list1 is None

def atIndex(lst: Optional[ListNode], index: int) -> Optional[ListNode]:
  if index < 0:
    return None
  for i in range(index):
    if lst is None:
      return None
    lst = lst.next
  return lst

File: Grind75/Utility/cpp/include/impl/List.cpp
#include <List.h>
#include <iostream>
#include <string>

namespace List {

  ListNode* makeList(std::vector<int> values) {
    ListNode *head = nullptr, *node = nullptr, *prev = nullptr;
    for (int v : values) {
      node = new ListNode(v);
      if (!head)
        head = node;
      if (prev)
        prev->next = node;
      prev = node;
    }
    return head;
  }

  void freeList(ListNode *lst) {
    ListNode *next;
    while (lst) {
      next = lst->next;
      delete lst;
      lst = next;
    }
  }

  void printList(ListNode *lst) {
    while (lst) {
      std::cout << lst->val << " -> ";
      lst = lst->next;
    }
    std::cout << '.' << std::endl;
  }

  std::string strList(ListNode *lst) {
    std::string output;
    while (lst) {
      output += std::to_string(lst->val) + " -> ";
      lst = lst->next;
    }
    output += '.';
    return output;
  }

  ListNode* tail(ListNode *head) {
    ListNode *node = head, *prev = nullptr;
    while (node) {
      prev = node;
      node = node->next;
    }
    return prev;
  }

  bool listEquals(ListNode *l1, ListNode *l2) {
    while (l1 && l2) {
      if (l1->val != l2->val)
        return false;
      l1 = l1->next;
      l2 = l2->next;
    }
    if (l1 || l2)
      return false;
    return true;
  }

  ListNode* atIndex(ListNode *lst, size_t index) {
    ListNode *node = lst;
    for (size_t i = 0; i < index; i++) {
      if (!node)
        return nullptr;
      node = node->next;
    }
    return node;
  }

}

